File: typescript/calculator.ts
// GitHub Copilot Tutorial - TypeScript Calculator
// Calculadora type-safe con manejo de errores

// Tipo union para operadores matemáticos válidos
type MathOperator = '+' | '-' | '*' | '/' | '%' | '^';

// Interface para el resultado de una operación
interface CalculationResult {
    result: number;
    error?: string;
    operation: string;
}

// Interface para el historial de operaciones
interface CalculationHistory {
    timestamp: Date;
    operation: string;
    result: number;
    inputs: number[];
}

// Type guard para validar números
function isValidNumber(value: unknown): value is number {
    // Validar que el valor sea un número válido y finito
    
}

// Type guard para validar operadores
function isValidOperator(operator: string): operator is MathOperator {
    // Validar que el operador esté en la lista de operadores válidos
    
}

// Clase Calculator con tipos estrictos
class TypeSafeCalculator {
    private history: CalculationHistory[] = [];
    private precision: number = 2;

    constructor(precision: number = 2) {
        // Inicializar calculadora con precisión específica
        
    }

    // Método genérico para operaciones básicas
    calculate(a: number, operator: MathOperator, b: number): CalculationResult {
        // Validar inputs usando type guards
        
        // Realizar operación según el operador
        
        // Agregar al historial
        
        // Retornar resultado formateado
        
    }

    // Método específico para suma con overloads
    add(a: number, b: number): number;
    add(numbers: number[]): number;
    add(aOrNumbers: number | number[], b?: number): number {
        // Implementar suma con múltiples signatures
        
    }

    // Método para operaciones avanzadas
    power(base: number, exponent: number): CalculationResult {
        // Calcular potencia con validaciones
        
    }

    // Método para raíz cuadrada
    sqrt(value: number): CalculationResult {
        // Calcular raíz cuadrada con validación de números negativos
        
    }

    // Método para factorial
    factorial(n: number): CalculationResult {
        // Calcular factorial con validación de enteros positivos
        
    }

    // Método para obtener historial tipado
    getHistory(): Readonly<CalculationHistory[]> {
        // Retornar copia readonly del historial
        
    }

    // Método para limpiar historial
    clearHistory(): void {
        // Limpiar array de historial
        
    }

    // Método para estadísticas del historial
    getStatistics(): {
        totalOperations: number;
        averageResult: number;
        mostUsedOperation: MathOperator | null;
        lastCalculation: CalculationHistory | null;
    } {
        // Calcular estadísticas del historial
        
    }

    // Método privado para agregar al historial
    private addToHistory(operation: string, result: number, inputs: number[]): void {
        // Agregar nueva entrada al historial
        
    }

    // Método privado para formatear resultado
    private formatResult(value: number): number {
        // Formatear resultado según precisión configurada
        
    }

    // Método para exportar historial como JSON
    exportHistory(): string {
        // Serializar historial a JSON
        
    }

    // Método para importar historial desde JSON
    importHistory(jsonData: string): boolean {
        // Deserializar y validar historial desde JSON
        
    }
}

// Función utilitaria para crear calculadora con configuración específica
function createCalculator(config: {
    precision?: number;
    maxHistorySize?: number;
}): TypeSafeCalculator {
    // Crear instancia de calculadora con configuración
    
}

// Función para validar expresión matemática como string
function parseExpression(expression: string): {
    isValid: boolean;
    tokens: (number | MathOperator)[];
    error?: string;
} {
    // Parsear y validar expresión matemática
    
}

// Función para evaluar expresión parseada
function evaluateExpression(tokens: (number | MathOperator)[]): CalculationResult {
    // Evaluar expresión usando orden de operaciones
    
}

// Clase para manejar conversiones de unidades
class UnitConverter {
    // Conversiones de temperatura
    static celsiusToFahrenheit(celsius: number): number {
        // Convertir Celsius a Fahrenheit
        
    }

    static fahrenheitToCelsius(fahrenheit: number): number {
        // Convertir Fahrenheit a Celsius
        
    }

    // Conversiones de distancia
    static metersToFeet(meters: number): number {
        // Convertir metros a pies
        
    }

    static feetToMeters(feet: number): number {
        // Convertir pies a metros
        
    }
}

// Ejemplo de uso
function demonstrateCalculator(): void {
    // Crear instancia de calculadora
    
    // Realizar operaciones básicas
    
    // Mostrar historial y estadísticas
    
    // Usar convertidor de unidades
    
}

// Ejecutar demostración si es archivo principal
if (require.main === module) {
    demonstrateCalculator();
}

export {
    TypeSafeCalculator,
    UnitConverter,
    createCalculator,
    parseExpression,
    evaluateExpression,
    isValidNumber,
    isValidOperator
};

export type {
    MathOperator,
    CalculationResult,
    CalculationHistory
};

File: rust/src/calculator.rs
// GitHub Copilot Tutorial - Rust Calculator CLI
// Calculadora robusta con manejo de errores y testing

use clap::{Parser, ValueEnum};
use anyhow::{Result, Context};
use thiserror::Error;
use std::fmt;

// Custom error types para la calculadora
#[derive(Error, Debug)]
pub enum CalculatorError {
    #[error("División por cero no permitida")]
    DivisionByZero,
    
    #[error("Operación no soportada: {operation}")]
    UnsupportedOperation { operation: String },
    
    #[error("Overflow en operación matemática")]
    Overflow,
    
    #[error("Entrada inválida: {input}")]
    InvalidInput { input: String },
}

// Enum para operaciones matemáticas
#[derive(Debug, Clone, ValueEnum)]
pub enum Operation {
    Add,
    Subtract,
    Multiply,
    Divide,
    Power,
    Sqrt,
    Factorial,
}

impl fmt::Display for Operation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Implementar Display para Operation
        
    }
}

// Struct para argumentos de línea de comandos
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
pub struct Args {
    /// Primera operando
    #[arg(short = 'a', long)]
    pub first: f64,
    
    /// Segunda operando (opcional para operaciones unarias)
    #[arg(short = 'b', long)]
    pub second: Option<f64>,
    
    /// Operación a realizar
    #[arg(short, long, value_enum)]
    pub operation: Operation,
    
    /// Precisión decimal para el resultado
    #[arg(short, long, default_value = "2")]
    pub precision: usize,
    
    /// Modo verbose para debugging
    #[arg(short, long)]
    pub verbose: bool,
}

// Struct principal de la calculadora
#[derive(Debug)]
pub struct Calculator {
    precision: usize,
    history: Vec<CalculationRecord>,
}

#[derive(Debug, Clone)]
pub struct CalculationRecord {
    operation: Operation,
    operands: Vec<f64>,
    result: f64,
    timestamp: std::time::SystemTime,
}

impl Calculator {
    // Constructor para nueva calculadora
    pub fn new(precision: usize) -> Self {
        // Crear nueva instancia con precisión especificada
        
    }

    // Método principal para realizar cálculos
    pub fn calculate(&mut self, operation: Operation, operands: &[f64]) -> Result<f64, CalculatorError> {
        // Validar operandos según la operación
        
        // Realizar cálculo usando match
        
        // Guardar en historial
        
        // Retornar resultado formateado
        
    }

    // Operaciones básicas
    fn add(&self, a: f64, b: f64) -> Result<f64, CalculatorError> {
        // Sumar dos números con check de overflow
        
    }

    fn subtract(&self, a: f64, b: f64) -> Result<f64, CalculatorError> {
        // Restar dos números
        
    }

    fn multiply(&self, a: f64, b: f64) -> Result<f64, CalculatorError> {
        // Multiplicar con check de overflow
        
    }

    fn divide(&self, a: f64, b: f64) -> Result<f64, CalculatorError> {
        // Dividir con validación de división por cero
        
    }

    fn power(&self, base: f64, exp: f64) -> Result<f64, CalculatorError> {
        // Calcular potencia
        
    }

    fn sqrt(&self, value: f64) -> Result<f64, CalculatorError> {
        // Raíz cuadrada con validación de números negativos
        
    }

    fn factorial(&self, n: f64) -> Result<f64, CalculatorError> {
        // Factorial con validación de enteros positivos
        
    }

    // Método para obtener historial
    pub fn get_history(&self) -> &[CalculationRecord] {
        // Retornar referencia al historial
        
    }

    // Método para limpiar historial
    pub fn clear_history(&mut self) {
        // Limpiar vector de historial
        
    }

    // Método para formatear resultado según precisión
    fn format_result(&self, value: f64) -> f64 {
        // Formatear con precisión especificada
        
    }

    // Método para validar operandos
    fn validate_operands(&self, operation: &Operation, operands: &[f64]) -> Result<(), CalculatorError> {
        // Validar número y tipo de operandos según operación
        
    }

    // Método para añadir al historial
    fn add_to_history(&mut self, operation: Operation, operands: Vec<f64>, result: f64) {
        // Crear registro y añadir al historial
        
    }
}

// Función principal del programa
fn main() -> Result<()> {
    // Parsear argumentos de línea de comandos
    let args = Args::parse();
    
    // Crear calculadora con precisión especificada
    
    // Preparar operandos según la operación
    
    // Realizar cálculo
    
    // Mostrar resultado
    
    // Si verbose, mostrar información adicional
    
}

// Función helper para mostrar información verbose
fn show_verbose_info(args: &Args, result: f64, calculator: &Calculator) {
    // Mostrar detalles de la operación y estado de la calculadora
    
}

// Tests unitarios
#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_basic_operations() {
        // Test operaciones básicas
        
    }
    
    #[test]
    fn test_error_conditions() {
        // Test condiciones de error
        
    }
    
    #[test]
    fn test_precision_formatting() {
        // Test formateo de precisión
        
    }
    
    #[test] 
    fn test_history_functionality() {
        // Test funcionalidad de historial
        
    }
}

// Property-based tests con quickcheck
#[cfg(test)]
mod property_tests {
    use super::*;
    use quickcheck_macros::quickcheck;
    
    #[quickcheck]
    fn addition_is_commutative(a: f64, b: f64) -> bool {
        // Test que la suma es conmutativa
        
    }
    
    #[quickcheck]
    fn multiplication_identity(a: f64) -> bool {
        // Test identidad multiplicativa
        
    }
}